crates/gosub_cairo/src/lib.rs

```diff
@@ -0,0 +1,134 @@
+use crate::elements::color::GsColor;
+use crate::elements::gradient::GsGradient;
+use crate::elements::image::GsImage;
+use crate::elements::rect::GsRect;
+use crate::elements::text::GsText;
+use crate::elements::transform::GsTransform;
+use crate::render::window::{ActiveWindowData, WindowData};
+use gosub_shared::render_backend::geo::SizeU32;
+use gosub_shared::render_backend::{RenderBackend, RenderRect, RenderText, Scene as _, WindowHandle};
+use gosub_shared::types::Result;
+pub use image::*;
+use log::info;
+pub use scene::*;
+use std::fmt::Debug;
+
+mod debug;
+mod elements;
+#[allow(unused)]
+pub mod render;
+mod scene;
+
+#[derive(Clone)]
+pub struct CairoBackend;
+
+impl Debug for CairoBackend {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("CairoRenderer").finish()
+    }
+}
+
+impl RenderBackend for CairoBackend {
+    type Rect = GsRect;
+    type Border = GsBorder;
+    type BorderSide = GsBorderSide;
+    type BorderRadius = GsBorderRadius;
+    type Transform = GsTransform;
+    type Text = GsText;
+    type Gradient = GsGradient;
+    type Color = GsColor;
+    type Image = GsImage;
+    type Brush = GsBrush;
+    type Scene = Scene;
+    type SVGRenderer = gosub_svg::resvg::Resvg;
+
+    type ActiveWindowData<'a> = ActiveWindowData;
+    type WindowData<'a> = WindowData;
+
+    fn draw_rect(&mut self, data: &mut Self::WindowData<'_>, rect: &RenderRect<Self>) {
+        data.scene.draw_rect(rect);
+    }
+
+    fn draw_text(&mut self, data: &mut Self::WindowData<'_>, text: &RenderText<Self>) {
+        data.scene.draw_text(text)
+    }
+
+    fn apply_scene(
+        &mut self,
+        data: &mut Self::WindowData<'_>,
+        scene: &Self::Scene,
+        transform: Option<Self::Transform>,
+    ) {
+        data.scene.apply_scene(scene, transform);
+    }
+
+    fn reset(&mut self, data: &mut Self::WindowData<'_>) {
+        data.scene.reset();
+    }
+
+    /// @todo: find out when this is called
+    fn activate_window<'a>(
+        &mut self,
+        _handle: impl WindowHandle + 'a,
+        data: &mut Self::WindowData<'_>,
+        _size: SizeU32,
+    ) -> Result<Self::ActiveWindowData<'a>> {
+        // I don't know what we need to do here. We have a handle (what does it hold?), the data (it
+        // holds the current main scene), and a size. It seems that the active window data just contains
+        // a cairo context. But we don't really need this in the active window data until we call
+        // self::render().. Maybe we get the context here, and not in the scene::render()?
+        Ok(ActiveWindowData {
+            cr: data._cr.as_ref().unwrap().clone(),
+        })
+    }
+
+    /// @todo: find out when this is called
+    fn suspend_window(
+        &mut self,
+        _handle: impl WindowHandle,
+        _data: &mut Self::ActiveWindowData<'_>,
+        _window_data: &mut Self::WindowData<'_>,
+    ) -> Result<()> {
+        Ok(())
+    }
```

Comment:
These two methods are called when the window is switching from shown to hidden or the other way around. I assume this is to save some memory by destroying the surfaces of the rendered window, which can be quite large in terms of memory.

I assume in cairo we can just ignore this, as we won't get this event by gtk and gtk should handle all of that internally, however this might not be always the case, for example when we use winit and vello

Reply:
i don't think this information should leak into the rendering backend to begin with.. for now, its fine as we don't use this in the cairo backend

Reply:
i guess we could change it, so the tab just renders to a scene and the rest is handled by the implementation, I mean in the future the tab doesn't get the full surface anyway

Reply:
yeah, there will be some more iterations on the rendering stuff later on.. at least we now have two different implementations so we can generalize the interface a bit better..